File: pa1/List.java
//---------------------------------------------------------------------------------------
//Name   : Anthony Lam
//Course : CMPS101
//Programming Assignment 1
//
//List.java
//This is a List ADT that incluedes a "cursor" that can be used for iteration. This List
//ADT uses a doubly linked list. Also contains a Node class.
//
// 
//---------------------------------------------------------------------------------------

public class List{

	//Node Class
	private class Node{
		//Fields or instance variables
		private Node next;
		private Node prev;
		private Node cursor;
		int data;

		public Node(int data){
			this.data = data;
			next = null;
			prev = null;
		}

		public String toString(){ //overrides original toString's method
			return String.valueOf(data);
		}

		public boolean equals(Object x){
			boolean equal = false;
			Node that;
			if(x instanceof Node){
				that = (Node) x;
				equal = (this.data == that.data);
			}
			return equal;
		}
	}

	//Beginning of List.java class
	//instance variables
private Node front;
private Node back;
private Node cursor;
private int length;
private int index;

//List Constructor
//Creates a new empty list.
public List() {
	Node front = null;
	Node back = null;
	cursor = null;
	index = -1;
	length = 0;
}

// Access functions----------------------------------------------------------------------

//int length()
//Returns the number of elements in this list.
public int length(){
	return length;
}

//int index()
//If cursor is defined, returns the index of the cursor element.
public int index(){
	if(this.length() <= 0){
		return -1;
	}
	return index;
}

//int front()
//Returns front element. Pre: length()>0
public int front(){
	if(this.length() <= 0){
		throw new RuntimeException(
			"List Error: front() called on an empty List");
	}
	else{
		return front.data;
	}
}

//int back()
//Returns back element. Pre:length()>0
public int back(){
	if (this.length() <= 0){
		throw new RuntimeException(
			"List Error: back() called on empty List");
	}
	else{
		return back.data;
	}
}

//int get()
//Returns cursor element. Pre: length() > 0, index() >= 0
public int get(){
	if(this.length() <= 0){
		throw new RuntimeException(
			"List error: get() called on an empty list.");
	}
	if(cursor == null){
		throw new RuntimeException(
			"List error: get() called with an undefined cursor.");
	}
	return cursor.data;
}

//boolean equals(List L)
//Returns true if and only if this List and L are the same
//integer sequence. The states of the cursors in the two Lists
//are not used in determining equality.
public boolean equals(List L){
	boolean eq = false;
	Node a, b;
	eq = (this.length() == L.length);
	if (eq == true){
		a = this.front;
		b = L.front;
		while(eq && a != null){
			eq = a.equals(b);
			a = a.next;
			b = b.next;
		}
	}
	return eq;
}


// Manipulation procedures---------------------------------------------------------------

//void clear()
//Resets this List to its original empty space.
public void clear() {
	this.front = null;
	this.back = null;
	index = -1;
	cursor = null;
	length = 0;

}

//void moveFront()
//If List is non-empty, places the cursor under the front element,
//otherwise do nothing.
public void moveFront(){
	if(this.length() > 0){
		cursor = this.front;
		index = 0;
	}
}

//void moveBack()
//If List is non-empty, places the cursor under the back element,
//otherwise does nothing.
public void moveBack(){
	if (this.length > 0){
		cursor = this.back;
		index = length - 1;
	}
}

//void movePrev()
//if cursor is defined and not at front, moves cursor one step toward
//undefined, if cursor is undefined does nothing.
public void movePrev(){
	if(cursor == front){
		cursor = null;
		index = -1;
	}
	else{
		cursor = cursor.prev;
		index--;
	}
}

//void moveNext()
//If cursor is defined and not at back, moves cursor one step toward
//undefined, if cursor is undefined does nothing.
public void moveNext(){
	if(cursor == back){
		cursor = null;
		index = -1;
	}
	else{
		cursor = cursor.next;
		index++;
	}
}

//void prepend(int data)
//Insert new element into this List. If List is non-empty,
//insertion takes place before front element.
public void prepend(int data){
	Node newNode = new Node(data);

	if(front == null){
		front = back = newNode;
		cursor = front;
	}
	else{
		this.front.prev = newNode;
		newNode.next = this.front;
		this.front = newNode;
		index++;
	}
	length++;
}


//void append(int data)
//Insert new element into this List. If List is non-empty,
//insertion takes place after back element.
public void append(int data){
	Node newNode = new Node(data);

	if(back == null){
		front = back = newNode;
		cursor = back;
	}
	else{
		this.back.next = newNode;
		newNode.prev = this.back;
		this.back = newNode;
		newNode.next = null;
	}
	length++;
}

//void insertBefore()
//Insert new element before cursor.
//Pre: length() > 0, index() >= 0
public void insertBefore(int data){
	Node newNode = new Node(data);

	if(this.length() <= 0){
		//throw new RuntimeException(
		//	"List Error: insertBefore() called with List empty.");
		prepend(data);
	}
	if(cursor == null){
		throw new RuntimeException(
			"List Error: insertBefore() called with cursor undefined.");
	}
	if(cursor == front){
		prepend(data);
	}
	else{
		newNode.next = cursor;
		newNode.prev = cursor.prev;
		cursor.prev.next = newNode;
		cursor.prev = newNode;
		index++;
		length++;
	}
}

//void insertAfter(int data)
//Inserts new element after cursor.
//Pre: length() > 0, index() > = 0
public void insertAfter(int data){
	Node newNode = new Node(data);

	if (this.length() <= 0){
		throw new RuntimeException(
			"List Error: insertAfter() called with List empty.");
		
	}
	if(this.index() < 0){
		throw new RuntimeException(
			"List Error: insertAfter() called with cursor undefined.");
	}
	if(cursor == back){
		append(data);
	}
	else{
		cursor.next.= newNode;
		newNode.prev = cursor;
		newNode.next = cursor.next; 
		cursor.next = newNode;
		length++;
	}
}

public void deleteFront(){
	if(this.length() <= 0){
		throw new RuntimeException(
			"List Error: deleteFront() called on a empty list.");
	}
	else{
		if(front == cursor){
			index = -1;
			cursor = null;
		}
		if(this.length() > 1){
			front = front.next;
			front.prev = null;
			
			if(cursor != null){
				index--;
			}
		}
		length--;
	}
}

public void deleteBack(){
	if(length() <= 0){
		throw new RuntimeException(
			"List Error: deleteBack() called on empty List.");
	}
	else{
		if(this.back == cursor){
			index = -1;
			cursor = null;
		}
		if(this.length() > 1){
			back = back.prev;
			back.next = null;
		}
		length--;
	}
}

public void delete(){
	if(this.length() <= 0){
		throw new RuntimeException(
			"List Error: delete() called on empty List.");
	}
	if(cursor == null){
		throw new RuntimeException(
			"List Error: delete() called with undefined cursor.");
	}
	if(cursor == front){
		deleteFront();
	}
	else if(cursor == back){
		deleteBack();
	}
	else{
		cursor.prev.next = cursor.next;
		cursor.next.prev = cursor.prev;
		length--;
	}
	index = -1;
}

//public String toString()
//Overides Object's toString method. Returns a String representation
//of this List consisting of a space sperated sequence of integers,
//with front on left.

public String toString(){
	Node newNode = front;
	StringBuffer s = new StringBuffer();

	while(newNode != null){
		s.append(newNode.data);
		if(newNode.next != null){
			s.append(" ");
		}
		newNode = newNode.next;
	}
	return new String(s);
}

//List copy()
//Returns a new List representing the same integer sequence as this
//List. The cursor in the new list is undefined, regardless of the
//state of the cursor in this List. This List is unchanged.

public List copy(){
	List l = new List();
	Node n = front;
	if(this.length() > 0){
		while(n != null){
			l.append(n.data);
			n = n.next;
		}
	}
	l.cursor = null;
	l.index = -1;
	return l;
}
}

File: pa5/List.c
//---------------------------------------------------------------------------------------
//Name  : Anthony Lam
//Course: CMPS101
//
//List.c
//Implementation file for List ADT
//
//---------------------------------------------------------------------------------------


#include<stdio.h>
#include<stdlib.h>
 
#include"List.h"

//structs--------------------------------------------------------------------------------

//private NodeObj type
typedef struct NodeObj{
	int length;
	struct NodeObj* next;
	struct NodeObj* prev;
	int data;
} NodeObj;

//private Node type
typedef NodeObj* Node;

//private ListObj type
typedef struct ListObj{
	Node front;
	Node back;
	Node cursor;
	int index;
	int length;
} ListObj;

//Constructors-Destructors---------------------------------------------------------------

//newNode()
//Returns reference to new Node object. Initializes next and data fields.
Node newNode(int data){
	Node N = malloc(sizeof(NodeObj));
	N -> data = data;
	N -> next = NULL;
	N -> prev = NULL;
	return (N);
}

//freeNode()
//Frees heap memory pointed to by *pN, sets *pN to NULL.
void freeNode(Node* pN){
	if(pN != NULL && *pN != NULL){
		free(*pN);
		*pN = NULL;
	}
}

//newList()
//Returns reference to new empty List object.
List newList(void){
	List L = malloc(sizeof(ListObj));
	L -> front = L -> back = NULL;
	L -> index = -1;
	L -> length = 0;
	return L;
}


//freeList()
//Frees all heap memory associated with List *pL, and sets *pL to NULL.S
void freeList(List* pL){
	if(pL != NULL && *pL != NULL){
		while(length(*pL) > 0){
			deleteFront(*pL);
		}
		free(*pL);
		*pL = NULL;
	}
}

//Access functions-----------------------------------------------------------------------

//getLength()
//Returns the length of L.
int length(List L){
	if(L == NULL){
		printf("List Error: calling getLength() on NULL List reference/n");
		exit(1);
	}
	return L->length;
}

//index(List L)
//Returns the index of the cursor in list.
int index(List L){
	if(L == NULL){
		printf("List Error: calling index() on NULL List reference.\n");
		exit(1);
	}
	if(length(L) <= 0){
		return -1;
	}
	return L -> index;
}

//front(List L)
//Returns the front element of the list.
int front(List L){
	if(L == NULL){
		printf("List Error: calling front(List L) on NULL list reference.\n");
		exit(1);
	}
	if(length(L) <= 0){
		printf("List Error: calling front(List L) on empty list.\n");
		exit(1);
	}
	return L->front->data;
}

//back(List L)
//Returns the back element of the list.
int back(List L){
	if(L == NULL){
		printf("List Error: calling back(List L) on NULL list reference.\n");
		exit(1);
	}
	if(length(L) <= 0){
		printf("List Error: calling back(List L) on empty list.\n");
		exit(1);
	}
	return L->back->data;
}

//get(List L)
//Returns the cursor element.
int get(List L){
	if(L == NULL){
		printf("List Error: calling get(List L) on a NULL list reference.\n");
		exit(1);
	}
	if(length(L) <= 0){
		printf("List Error: calling get(List L) on a empty list.\n");
		exit(1);
	}
	if( L -> cursor == NULL){
		printf("List Error: calling get(List L) on NULL cursor.\n");
		exit(1);
	}
	return L-> cursor -> data;
}

//equals(List A, List B)
//Returns true if and only if List A and B are the same integer sequence.
//The states of the cursors in the two lists are not used in determining equality.

int equals(List A, List B){
	int eq = 0;
	if(A == NULL|| B == NULL){
		printf("List Error: calling equals(List A, List B) with at least one NULL list reference.\n");
		exit(1);
	}

	eq = (A -> length == B -> length);
	Node M = A -> front;
	Node N = B -> front;
	
	while( eq && N != NULL){
		eq = (N -> data == M -> data);
		N = N -> next;
		M = M -> next;
	}
	return eq;
}

//Manipulation procedures----------------------------------------------------------------

//clear(List L)
//Clears List L to empty state.
void clear(List L){
	if(L == NULL){
		printf("List Error: calling clear(List L) on NULL list reference.\n");
		exit(1);
	}

	while(length(L) > 0){
		deleteFront(L);
	}
	L-> length = 0;
	L-> index = -1;
	L-> front = L-> back = L -> cursor = NULL;
}

//moveFront(List L)
//Moves cursor to the front.
void moveFront(List L){
	if(L == NULL){
		printf("List Error: calling moveFront(List L) on NULL list reference.\n");
		exit(1);
	}
	if(length(L) > 0){
		L -> cursor = L -> front;
		L -> index = 0;
	}
}

//moveBack(List L)
//Moves cursor to the back.
void moveBack(List L){
	if(L == NULL){
		printf("List Error: calling moveBack(List L) on NULL list reference.\n");
		exit(1);
	}
	if(length(L) > 0){
		L -> cursor = L -> back;
		L -> index = (L -> length) - 1;
	}
}

//moveNext(List L)
//Moves cursor to next node of current Node.
void moveNext(List L){
	if(L == NULL){
		printf("List Error: calling moveNext(List L) on NULL list reference.\n");
	}
	if(L -> cursor == L -> back){
		L -> cursor = NULL;
		L -> index = -1;
	}
	else{
		L -> cursor = L -> cursor -> next;
		L -> index++;
	}
}

//movePrev(List L)
//Moves cursor to previous node of current Node.
void movePrev(List L){
	if(L == NULL){
		printf("List Error: calling movePrev(List L) on NULL list reference.\n");
		exit(1);
	}
	if(L -> cursor == L -> front){
		L -> cursor = NULL;
		L -> index = -1;
	}
	else{
		L -> cursor = L -> cursor -> prev;
		L -> index--;
	}
}

//prepend(List L, int data)
//Inserts new element into the list, if not empty, inserts element
//before the front element.
void prepend(List L, int data){
	if(L == NULL){
		printf("List Error: calling prepend(List L, int data) on NULL list reference.\n");
		exit(1);
	}

	

	if(L -> front == NULL){
		Node N = newNode(data);
		L -> front = L -> back = N;
		L -> cursor = L -> front;
	}
	else{
		Node N = newNode(data);
		L -> front -> prev = N;
		N -> next = L -> front;
		L -> front = N;
		L -> index++;
	}
	L -> length++;
}

//append(List L, int data)
//Inserts new element into the list, if not empty, inserts element
//after the back element.
void append(List L, int data){
	if(L == NULL){
		printf("List Error: calling append(ListL, int data) on NULL list reference.\n");
		exit(1);
	}

	if(L -> back == NULL){
		Node N = newNode(data);
		L -> front = L -> back = N;
		L -> cursor = L -> back;
	}
	else{
		Node N = newNode(data);
		L -> back -> next = N;
		N -> prev = L -> back;
		L -> back = N;
		N -> next = NULL;
	}
	L -> length++;
}

//insertBefore(List L, int data)
//Inserts new element before cursor.
void insertBefore(List L, int data){
	if(L == NULL){
		printf("List Error: calling insertBefore(List L, int data) on NULL list reference.\n");
		exit(1);

	}
	if(L -> cursor == NULL){
		printf("List Error: calling insertBefore(List L, int data) with NULL cursor.\n");
		exit(1);
	}
	
	if(L -> length <= 0){
		prepend(L, data);
	}
	if(L -> cursor == L -> front){
		prepend(L, data);
	}
	else{
		Node N = newNode(data);
		N -> next = L -> cursor;
		N -> prev = L -> cursor -> prev;
		L -> cursor -> prev -> next = N;
		L -> cursor -> prev = N;
		L -> index++;
		L -> length++;
	}
}

//insertAfter(List L, int data)
//Inserts new element after cursor.
void insertAfter(List L, int data){
	if(L == NULL){
		printf("List Error: calling insertAfter(List L, int data) on NULL list reference.\n");
		exit(1);
	}
	if(L -> cursor == NULL){
		printf("List Error: calling insertAfter(List L, int data) with NULL cursor.\n");
		exit(1);
	}

	if(L -> cursor == L -> back){
		append(L, data);
	}
	else{
		Node N = newNode(data);
		L -> cursor -> next -> prev = N;
		N -> prev = L -> cursor;
		N -> next = L -> cursor -> next;
		L -> cursor -> next = N;
		L -> length++;
	}
}

//deleteFront(List L)
//Deletes element at the front of the list L.
void deleteFront(List L){
	if(L == NULL){
		printf("List Error: calling deleteFront(List L) on NULL list reference.\n");
		exit(1);
	}
	if(L -> length <= 0){
		printf("List Error: calling deleteFront(List L) on empty list.\n");
		exit(1);
	}
	else{
			if(L-> length == 1){
				Node N = L -> front;
				L -> front = L -> back = NULL;
				freeNode(&N);
				L -> index = -1;
				L -> cursor = NULL;
			}
			else{
				Node N = L -> front;
				L -> front = L -> front -> next;
				L -> front -> prev = NULL;
				if(L -> cursor != NULL){
					L -> index--;
				}
				freeNode(&N);
			}
			//freeNode(&N);
			L -> length--;
	}
	
}

//deleteBack(List L)
//Delete element at the back of the list L.
void deleteBack(List L){
	if(L ==NULL){
		printf("List Error: calling deleteFront(List L) on NULL list reference.\n");
		exit(1);
	}
	if(L -> length <= 0){
		printf("List Error: calling deleteFront(List L) on empty list.\n");
		exit(1);
	}
	else{
			if(L -> length == 1){
				Node N = L -> back;
				L -> front = L -> back = NULL;
				freeNode(&N);
				L -> index = -1;
				L -> cursor = NULL;
				
			}
			else{
				Node N = L -> back;
				L -> back = L -> back -> prev;
				L -> back -> next = NULL;
				if(L -> index == L -> length -1){
					L -> index = -1;
				}
				freeNode(&N);
			}
			//freeNode(&N);
			L -> length--;
		}		
}



//delete(List L)
//Deletes element at cursor element.
void delete(List L){
	if(L == NULL){
		printf("List Error: calling delete(List L) on NULL list reference.\n");
		exit(1);
	}
	if(L -> length <= 0){
		printf("List Error: calling delete(List L) on empty list.\n");
		exit(1);
	}
	if(L -> cursor == L -> front){
		deleteFront(L);
	}
	else if(L -> cursor == L -> back){
		deleteBack(L);
	}
	else{
		Node N = L -> cursor;
		L -> cursor -> prev -> next = L -> cursor -> next;
		L -> cursor -> next -> prev = L -> cursor -> prev;
		L -> cursor = NULL;
		freeNode(&N);
		L -> length--;
		
	}
	
	L -> index =-1;	
}



//Other operations-----------------------------------------------------------------------


//printList(FILE* out, List L)
//Returns List as a string representation.
void printList(FILE* out, List L){
	if(L == NULL){
		printf("List Error: calling printList(FILE* out, List L) on NULL list reference.\n");
		exit(1);
	}
	Node N = L -> front;
	while(N != NULL){
		fprintf(out, "%d ", N -> data);
		N = N -> next;
	}
	printf("\n");
}

//List copyList(List L)
//Makes a new list and copies L to the new list.
List copyList(List L){
	Node N = L -> front;
	List M = newList();
	if(L -> length > 0){
		while(N != NULL){
			append(M, N -> data);
			N = N -> next;
		}
	}
	M -> cursor = NULL;
	M -> index = -1;
	return M;
}



/* deleteback
Node N = L -> back;
		if( L -> length == 1){
			L -> front = L -> back = NULL;
			L -> index = -1;
			L -> cursor = NULL;
		}
		else{
			if(L -> cursor == L -> back){
				L -> index = -1;
			}

			L -> back = L -> back -> prev;
			L -> back -> next = NULL;


deletefront
Node N = L -> front;
		if(L -> length == 1){
			L -> front = L -> back = NULL;
			L -> index = -1;
			L -> cursor = NULL;
		}
		else{
			if(L -> cursor != NULL){
				L -> index--;
			}

			L -> front = L -> front -> next;
			L -> front -> prev = NULL;
*/